### PR TITLE
feat: Allow using binary for document extraction

### PR DESCRIPTION
**Pull request by Ngkaokis**
ref #2609

### CODE REVIEW

**Comment by roxk** [comment]
Still reviewing, but have you added the checklist.md yet? Please add "Make sure private image works" to the checklist.

**Comment by Ngkaokis** [comment]
> Still reviewing, but have you added the checklist.md yet? Please add "Make sure private image works" to the checklist.

Not yet. Let me add it now

File: nodes/FormX/actions/document/commonProperties.ts [hunk 1]
```diff
@@ -4,11 +4,13 @@ export const commonProperties = (
...
 ): INodeProperties[] => [
 	{
-		displayName: 'Image URL',
-		name: 'imageUrl',
+		displayName: 'Binary Data Field',
+		name: 'binaryDataField',
 		type: 'string',
-		default: '',
-		placeholder: 'https://formextractorai.com/sample-invoice-1.d551279a.jpg',
+		default: 'data',
```
Comment (roxk) [inline nodes/FormX/actions/document/commonProperties.ts:10]: I think we should support both image URL or binary file...? Or does n8n support uploading the file directly?

**Review by roxk — COMMENTED** [review]
I know you are fixing n8n only, but could you make it so that the api client between zapier and this repo is the same? We might need to come up with a way to make the api client code shared (another sub-module maybe...)

File: nodes/FormX/actions/document/commonProperties.ts (hunk 1 above)
Comment (Ngkaokis) [reply nodes/FormX/actions/document/commonProperties.ts:10]: Yes, both are supported. But by default binary file is expected.

If want to use image url, can pass the url in additional field

Comment (roxk) [reply nodes/FormX/actions/document/commonProperties.ts:10]: Gotcha

**Comment by roxk** [comment]
Btw I found CI is still not setup 🙈

**Comment by Ngkaokis** [comment]
> Btw I found CI is still not setup 🙈

Yes, will do after declarative style refactoring as quite a lot of code change there so want to do it first.

File: nodes/FormX/actions/document/asyncExtract.operation.ts [hunk 2]
```diff
@@ -23,19 +23,22 @@ const displayOptions: IDisplayOptions = {
 export const description = updateDisplayOptions(displayOptions, properties);
 
 export async function execute(this: IExecuteFunctions, i: number): Promise<INodeExecutionData[]> {
-	const imageUrl = this.getNodeParameter('imageUrl', i, undefined, {
-		extractValue: true,
-	});
+	const binaryDataField = this.getNodeParameter('binaryDataField', i);
+
+	let dataBuffer: Buffer;
```
Comment (roxk) [inline nodes/FormX/actions/document/asyncExtract.operation.ts:28]: Consider adding a func that get binary data field in 1 line as I see this code repeated several times.

**Review by roxk — CHANGES_REQUESTED** [review]

**Comment by roxk** [comment]
> I suspect whole api client will be removed in n8n case.

I can't imagine how that works but certainly could use a follow up issue for making sure api client used across repos is consistent. Please add an issue for that, thanks!

**Comment by Ngkaokis** [comment]
@roxk Updated! THX!

>I can't imagine how that works but certainly could use a follow up issue for making sure api client used across repos is consistent. Please add an issue for that, thanks!

Follow up issue: https://linear.app/formx/issue/FORMX-2623/n8n-investigate-how-to-extract-api-client-as-common-package

**Review by roxk — APPROVED** [review]

**Comment by roxk** [comment]
@Ngkaokis Approved. I dont have write access again for sone reason so pls merge yourself first. If you are the admin please grant me access, thanks!

**Pull request merged**